### PR TITLE
feat: add support for S3 compatible storage

### PR DESCRIPTION
**Pull request by Kare-Udon**
Add a new storage option: S3.
You can use any S3 compatible services.

### CODE REVIEW

File: nazurin/storage/s3.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+    async def check_bucket(self):
+        if not S3.client.bucket_exists(S3_BUCKET):
+            S3.client.make_bucket(S3_BUCKET)
+            logger.info("Bucket created: {}", S3_BUCKET)
+
+    async def upload(self, file: File):
+        S3.client.fput_object(
+            bucket_name=S3_BUCKET,
+            object_name=file.name,
```
Comment (y-young) [inline nazurin/storage/s3.py:39]: Concatenate `file.destination` and `file.name` to build the path.

File: docs/includes/storage.zh.md [hunk 2]
```diff
@@ -5,3 +5,4 @@
 |                                [MEGA](https://mega.io/)                                 |         [MEGA](/zh/storage/mega)         |    `Mega`     |              |
 |                        [Google Drive](https://drive.google.com/)                        | [Google Drive](/zh/storage/google-drive) | `GoogleDrive` |              |
 | [OneDrive](https://www.microsoft.com/zh-cn/microsoft-365/onedrive/online-cloud-storage) |     [OneDrive](/zh/storage/onedrive)     |  `OneDrive`   |              |
+| [S3](https://aws.amazon.com/s3/)                                                         |         [S3](/storage/s3)             |     `S3`      |             |
```
Comment (y-young) [inline docs/includes/storage.zh.md:8]: `/zh/storage/s3`

File: nazurin/storage/s3.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
+# -*- coding: utf-8 -*-
```
Comment (y-young) [inline nazurin/storage/s3.py:1]: No longer needed.

File: nazurin/storage/s3.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+from nazurin.config import env
+from nazurin.models import File
+from nazurin.utils import logger
+
+S3_ENDPOINT = env.str("S3_ENDPOINT", default="s3.amazonaws.com")
+S3_ACCESS_KEY = env.str("S3_ACCESS_KEY")
+S3_SECRET_KEY = env.str("S3_SECRET_KEY")
+S3_SECURE = env.bool("S3_SECURE", default=True)
+S3_REGION = env.str("S3_REGION", default="null")
```
Comment (y-young) [inline nazurin/storage/s3.py:16]: Are you sure it's not `None`?

File: nazurin/storage/s3.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        region=S3_REGION,
+        secure=S3_SECURE,
+    )
+
+    async def check_bucket(self):
+        if not S3.client.bucket_exists(S3_BUCKET):
+            S3.client.make_bucket(S3_BUCKET)
+            logger.info("Bucket created: {}", S3_BUCKET)
+
+    async def upload(self, file: File):
```
Comment (y-young) [inline nazurin/storage/s3.py:36]: Seems the client is not asynchronous, try using `async_wrap`, same for other methods.

File: nazurin/storage/s3.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+import mimetypes
+from typing import List
+
+from minio import Minio
+
+from nazurin.config import env
+from nazurin.models import File
+from nazurin.utils import logger
+
+S3_ENDPOINT = env.str("S3_ENDPOINT", default="s3.amazonaws.com")
```
Comment (y-young) [inline nazurin/storage/s3.py:12]: Use `with env.prefixed("S3_"):`

**Review by y-young — CHANGES_REQUESTED** [review]

File: nazurin/storage/s3.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    def check_bucket(self):
+        if not S3.client.bucket_exists(BUCKET):
+            S3.client.make_bucket(BUCKET)
+            logger.info("Bucket created: {}", BUCKET)
+
+    @async_wrap
+    def upload(self, file: File):
+        S3.client.fput_object(
+            bucket_name=BUCKET,
+            object_name=f"{file.destination}/{file.name}",
```
Comment (y-young) [inline nazurin/storage/s3.py:42]: Use `pathlib.Path` and `as_posix()`

Comment (Kare-Udon) [reply nazurin/storage/s3.py:42]: sry, typical error :(

**Pull request merged**